### DISCUSSION TITLE
Key Rotation

### DISCUSSION THREAD

**Discussion by cjc166** [NetSparkleUpdater/NetSparkle#333] [General]
Does anyone have any suggestions/insight into how to perform key rotation for the Ed25519 keys that are used for the app.....for security purposes I'd like to be able to rotate the private key that I use for signatures, but that would pose issues to applications out there with the existing private key...curious if anyone has tips or insight here or if this is supported?

**Comment by Deadpikle** [comment]
Hi @cjc166, sorry for missing this.

Please note that the private key should NOT shipped with your app in any way, shape, or form. You should only ship your public key. If you are shipping your private key with your application, that is a security vulnerability created by, er, you.

There's not really a great way to do key rotation currently. You'd have to create a new public/private key pair, send out an update to all your users that was signed with the old pair. The new software update would have your new public key. You'd probably also want to swap app cast URLs to keep things simple and make sure old clients don't see updates that are signed with the new key pair. Then future updates would use the new key pair and you'd sign all future updates / app casts with your new key pair. This is the same process as upgrading from DSA keys to Ed25519 for this lib. Since you're replacing the key, your current software doesn't know about the current key, so you've gotta provide a software update (signed with the key pair that it understands/knows) in order to get the new key out there. At least, that's how it works currently.
